infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/NamespaceNameValidator.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.namespace;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.lang.String.format;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.regex.Pattern;
+import org.eclipse.che.api.core.ValidationException;
+
+public final class NamespaceNameValidator {
```

Comment:
I don't understand this class. You're using result object inside for some reason, but resulting in exceptions on public API of util class. Why?
I hate these validate methods throwing exceptions, because it leads to controlling flow with exceptions which is nasty. But at least make it simple. Why not throw these exceptions in `validateInternal`? which should be private.

Reply:
It should not be util class.

Reply:
This class is a utility class because it is used a) in a workspace attribute validator (and having multi-level validation just because of the namespaces is IMHO an utter overkill) and b) in methods that are outside of validator workflow and operate with information outside of validator's reach.

There are 2 ways of using this utility class. Either you want to report on error in the name (e.g. in the validator workflow) or you want to just check that the name is valid (hence, the 2 public methods).

The internal enum is used to efficiently implement these two usecases without duplicating logic. 

The `validateInternal` method is not private because it is used in tests that would otherwise rely on exception message wording with I think is very brittle.